main/src/library/BufReader.flix

```diff
@@ -0,0 +1,183 @@
+/*
+ *  Copyright 2025 Ry Wiese
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+///
+/// A `Reader` implementation that wraps an underlying `Reader` instance
+/// and caches the bytes read from the underlying instance in an internal buffer.
+///
+/// A call to `read` will read directly from the buffer if bytes are available, and
+/// will trigger a `read` from the underlying `reader` when the buffer is empty.
+///
+/// This has the effect of reducing the overall number of `read`s to the underlying
+/// `Reader`.
+///
+/// A call to `read` might read more than the requested number of bytes from the
+/// underlying reader, in order to fill the entire buffer.
+///
+/// A call to `read` might not read from the underlying `reader` at all, if there are
+/// enough bytes in the buffer.
+///
+pub struct BufReader[t: Type, r: Region] {
+    buffer: Array[Int8, r],
+    mut cursor: Int32,
+    mut size: Int32,
+    reader: t
+}
+
+instance Reader[BufReader[t, rr]] with Reader[t] {
+
+    type Aef = rr + Reader.Aef[t]
+
+    pub def read(
+        buffer: Array[Int8, rb],
+        bufReader: BufReader[t, rr]
+    ): Result[IoError, Int32] \ rb + rr + Reader.Aef[t] =
+        BufReader.read(buffer, bufReader)
+
+}
+
+mod BufReader {
+
+    ///
+    /// Constructs a `BufReader` wrapping `reader` with a default capacity of 4096 bytes.
+    ///
+    pub def withDefaultCapacity(
+        rc: Region[r],
+        reader: t
+    ): BufReader[t, r] \ r =
+        withCustomCapacity(rc, defaultBufferCapacity(), reader)
+
+    ///
+    /// Constructs a `BufReader` wrapping `reader` with the specified `capacity` in bytes.
+    ///
+    /// If `capacity <= 0`, the `BufReader` will delegate directly to `reader` without buffering.
+    ///
+    pub def withCustomCapacity(
+        rc: Region[r],
+        capacity: Int32,
+        reader: t
+    ): BufReader[t, r] \ r =
+        new BufReader @ rc {
+            buffer = Array.repeat(rc, capacity, 0i8),
+            cursor = 0,
+            size = 0,
+            reader = reader
+        }
```

Comment:
Perhaps `emptyWithDefaultCapacity` and `emptyWithCapacity` are better?

Reply:
rust has `new` and `with_capacity`

Reply:
Should I use those?

Reply:
Might be hard to use `new` since that's a keyword I think

Reply:
Yeah, it is a keyword